drivers/gpio/gpio_altera_pio.c

```diff
@@ -0,0 +1,293 @@
+/*
+ * Copyright (c) 2023, Intel Corporation. All rights reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT altr_pio_1_0
+
+#include <zephyr/device.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/gpio/gpio_utils.h>
+
+#define ALTERA_AVALON_PIO_DATA_OFFSET		0x00
+#define ALTERA_AVALON_PIO_DIRECTION_OFFSET	0x04
+#define ALTERA_AVALON_PIO_IRQ_OFFSET		0x08
+#define ALTERA_AVALON_PIO_SET_BITS		0x10
+#define ALTERA_AVALON_PIO_CLEAR_BITS		0x14
+
+typedef void (*altera_cfg_func_t)(void);
+
+struct gpio_altera_config {
+	struct gpio_driver_config	common;
+	uintptr_t			reg_base;
+	uint32_t			irq_num;
+	uint8_t				direction;
+	altera_cfg_func_t		cfg_func;
+};
+
+struct gpio_altera_data {
+	/* gpio_driver_data needs to be first */
+	struct gpio_driver_data common;
+	/* list of callbacks */
+	sys_slist_t cb;
+};
+
+static bool gpio_pin_direction(const struct device *dev, uint32_t pin_mask)
+{
+	const struct gpio_altera_config *cfg = dev->config;
+	const int direction = cfg->direction;
+	uintptr_t reg_base = cfg->reg_base;
+	uint32_t addr;
+	uint32_t pin_direction;
+
+	if (pin_mask == 0) {
+		return -EINVAL;
+	}
+
+	/* Check if the direction is Bidirectional */
+	if (direction != 0) {
+		return -EINVAL;
+	}
+
+	addr = reg_base + ALTERA_AVALON_PIO_DIRECTION_OFFSET;
+
+	pin_direction = sys_read32(addr);
+
+	if (!(pin_direction & pin_mask)) {
+		return false;
+	}
+
+	return true;
+}
+
+static int gpio_altera_configure(const struct device *dev,
+				 gpio_pin_t pin, gpio_flags_t flags)
+{
+	const struct gpio_altera_config *cfg = dev->config;
+	const int port_pin_mask = cfg->common.port_pin_mask;
+	const int direction = cfg->direction;
+	uintptr_t reg_base = cfg->reg_base;
+	uint32_t addr;
+
+	/* Check if pin number is within range */
+	if ((port_pin_mask & BIT(pin)) == 0) {
+		return -EINVAL;
+	}
+
+	/* Check if the direction is Bidirectional */
+	if (direction != 0) {
+		return -EINVAL;
+	}
+
+	addr = reg_base + ALTERA_AVALON_PIO_DIRECTION_OFFSET;
+
+	if (flags == GPIO_INPUT) {
+		sys_clear_bits(addr, BIT(pin));
+	} else if (flags == GPIO_OUTPUT) {
+		sys_set_bits(addr, BIT(pin));
+	} else {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int gpio_altera_port_get_raw(const struct device *dev, uint32_t *value)
+{
+	const struct gpio_altera_config *cfg = dev->config;
+	uintptr_t reg_base = cfg->reg_base;
+	uint32_t addr;
+
+	addr = reg_base + ALTERA_AVALON_PIO_DATA_OFFSET;
+
+	__ASSERT(value != NULL, "value is null pointer!");
+
+	if	(value == NULL)	{
```

Comment:
You have a number of lines in this file with excessive whitespace. I suggest running `clang-format` on it.

Also, no need for both an assert and if statement.

Reply:
The "if" statement is used to prevent the null pointer. Does the "if" statement need to be removed?